### PR TITLE
BIC-212 # Fix scrollbar dissapearing

### PR DESCRIPTION
**Pull request by simonmarklar**
- Changed animationfinish event to listen on the window element

### CODE REVIEW

File: src/bic/fix-animation.js [hunk 1]
```diff
@@ -26,15 +27,17 @@ define(function (require) {
...
 
-  $(window).on('animationstart', function (event) {
+  $window.on('animationstart', function (event) {
     var timer;
     if (isJQMTransition(event)) {
       // okay, we need to set a 1s timer in case this doesn't finish
       timer = setTimeout(function () {
         c.warn('"animationend" event expected but not observed');
-        $(event.target).trigger('animationend');
+        $window.trigger('animationend');
```
Comment (jokeyrhyme) [inline src/bic/fix-animation.js]: Assuming the event bubbles properly, I wonder if it wouldn't be safer to still `.trigger()` the event on the original target element?

**Comment by simonmarklar** [comment]
Some notes:

- Blackboxing both jquery and jquery mobile in chrome fixed the error
- When stepping through a click event, none of our code is run
- Firefox runs the animationend event whilst no other browser does
- Someone has reported an issue  https://github.com/jquery/jquery-mobile/issues/6515 that matches this, but is unable to reproduce
- When making a plain html that just loads jquery + jquery mobile with a enough text to need a scroll bar and a footer element, the issue does not happen
- adding a button which swaps out html to the previous step and triggers `pagecreate` doesn't trigger the error

**Comment by jokeyrhyme** [comment]
I'm wondering if this code makes it difficult to have overlapping animations. e.g:
1. 0s: A starts on element A
2. 1s: B starts on element B
3. 2s: A ends on element A
4. 3s: B ends on element B

This sort of thing seems like it would be hard to achieve once the element information is discarded.

Although, there's probably nothing in jQueryMobile that constructs such animation sets in the first place. And I wonder how many customers would want to do this?

File: src/bic/fix-animation.js (hunk 1 above)
Comment (simonmarklar) [inline src/bic/fix-animation.js]: aaaah!!!!

i missed one simple thing when reading the animation docs:

target: document, element.

in other words, it doesn't bubble up to the window  where the listener is!

**Comment by jokeyrhyme** [comment]
Ah, I think listening on `window` was my bad, although we had this issue long before I wrote this watchdog.

LGTM :+1:

**Comment by simonmarklar** [comment]
Yeah that sequence of events might be problematic, I'll add some a counter to make sure that the class is only removed when all animations have finished, just to be sure.

I think all I've done here is hang a painting over a hole. But i have a bad feeling to find the hole we'll have to strip back everything and add them one by one.

I've got a test setup with just the basics (i.e. no backbone and no bic), and the issue cannot be reproduced using the same steps. I thought it was  way we set the views using .html() to render the page rather than just updating the header, content and footer as the jquery docs suggest, but that works fine (we probably should switch from render all + enhancing to just render part + refresh, but thats another ticket). I stopped there because any further coding would mean pretty much writing bic again. As much as I want to do that.... ;)

I bet there's something that was done, very early on in the project, to get something to work that has this unintended side effect.

**Comment by jokeyrhyme** [comment]
Oh, I thought the scenario I posed was actually already satisfied when you reverted the change to the `.trigger()` line. Is the counter necessary?

**Comment by simonmarklar** [comment]
yeah i think the counter is necessary. What you proposed made me think "what if some fancy page change multi-animation is running, or what if the page change animation is queued up to run after an element slide event, what if they open a side bar and tap on multiple interaction tags triggering multiple page change events??"

When you turn debug on and click the screen really fast, you see
there are n animations running
...
there are 2 animations running
there are 1 animations running
making sure the transitioning classes are being removed.

so if ever there is more than one animation running, the transition classes should be removed only when the final animation runs. if they've already been removed, then nothing will happen. 

This is just a catch all because I dont think that finding the root cause is going to be simple. I reduced the problem down to the core components and it all worked as expected, the only thing missing was bic...

**Comment by jokeyrhyme** [comment]
Would it be better to split the counters and class fix into a separate set of event handlers, so that they are separate from the watchdog? This is rough, but maybe something like:

```js
$document.on('animationstart', function () { /* TODO: original watchdog */ });

$document.on('animationstart', function () { counter++; });
$document.on('animationend', function () { counter--; /* TODO: fix class if <= 0 */ });
```

That `$document.one()` was from the original watchdog, and it's not clear to me whether it's a perfect match for the `$document.on()` in this particular case. Maybe it is, I'm probably just having a caffeine-deprived moment. /shrug

**Comment by jokeyrhyme** [comment]
You seem confident IRL, and this _does_ look good enough to me. At this stage I'm satisfied my questions have been considered. :+1: :smile_cat:

**Comment by simonmarklar** [comment]
yeah sorry my head is buried in windows and visual studios constant crashing. Its frustrating the hell out of me

ux review from ray.

tested in windows:
- Chrome
- IE11
- Edge
- FF (but FF worked originally, which lead me to the problem)
- Windows App

Android
- Chrome
- mBaaS trial app

**Pull request merged**